File: src/aoxcon_ref/state_machine.py
from dataclasses import dataclass
from .errors import ERR_INVALID_STATE_TRANSITION

VALID_TRANSITIONS = {
    "Initialized": {"Queued"},
    "Queued": {"Verified", "Rejected"},
    "Verified": {"Executed", "FailedExecution"},
    "Executed": {"Finalized"},
    "Finalized": set(),
    "Rejected": set(),
    "FailedExecution": set(),
}


@dataclass(frozen=True)
class TransitionResult:
    ok: bool
    code: str
    from_state: str
    to_state: str


def transition(from_state: str, to_state: str) -> TransitionResult:
    if to_state in VALID_TRANSITIONS.get(from_state, set()):
        return TransitionResult(True, "OK_PROCESSED", from_state, to_state)
    return TransitionResult(False, ERR_INVALID_STATE_TRANSITION, from_state, to_state)

File: haskell/aoxcon-cardano-ref/src/Aoxcon/StateMachine.hs
module Aoxcon.StateMachine where

import Aoxcon.Errors (errInvalidStateTransition, okProcessed)
import Aoxcon.Types (TransitionResult(..))
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set

validTransitions :: Map.Map String (Set.Set String)
validTransitions =
  Map.fromList
    [ ("Initialized", Set.fromList ["Queued"])
    , ("Queued", Set.fromList ["Verified", "Rejected"])
    , ("Verified", Set.fromList ["Executed", "FailedExecution"])
    , ("Executed", Set.fromList ["Finalized"])
    , ("Finalized", Set.empty)
    , ("Rejected", Set.empty)
    , ("FailedExecution", Set.empty)
    ]

transition :: String -> String -> TransitionResult
transition fromS toS =
  let allowed = Map.findWithDefault Set.empty fromS validTransitions
   in if Set.member toS allowed
        then TransitionResult True okProcessed fromS toS
        else TransitionResult False errInvalidStateTransition fromS toS
